Re-write initcluster.vbs into powershell script and fix the initdb failure when the locale passed through the installer script contains non-ascii characters. (v17)

Manika-EDB opened the pull request:
The commit handles two issues in a single commit:

With Microsoft announcing the retirement of VBScript, this commit replaces the
last used VB script (initcluster.vbs) to a powershell script (initcluster.ps1).

It also addresses the issue of database initialisation failure if locale contains
non-ASCII characters. During installation, the EDB GUI installer would run
getlocales.exe retrieving a list of all the available system locales, populate the
list in a drop down box for users to choose the desired locale. The locale
containing the non-ascii characters when passed to initdb.exe returns the
error. This behaviour changes with this commit

Replace getlocales.exe with a powershell command to return the list of
locales. When user chooses a locale, it'll now be converted into BCP-47 code
name before passing to initcluster script and further to initdb.exe.

DBP-719, DBP-788, 
edb-installers/issues - #127, #148, #186, #193, #220
--Manika Singhal, Reviewed by- Dave Page, Sandeep Thakkar, Tested By- Manika Singhal, Kashif Zeeshan

Manika-EDB committed:
testing initcluster.ps1

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

dpage commented on server/scripts/windows/initcluster.ps1:
```diff
@@ -0,0 +1,321 @@
+# PowerShell Script for PostgreSQL Cluster Initialization
+# Copyright (c) 2012-2022, EnterpriseDB Corporation.  All rights reserved
```
2025?

Manika-EDB replied:
fixed it.

dpage commented on server/scripts/windows/initcluster.ps1:
```diff
@@ -0,0 +1,321 @@
+Write-Host "`nInitializing PostgreSQL database cluster..."
+$iRet = DoCmd -Command "$initdbCmd"
+if ($iRet -ne 0) {
+    Die "Failed to initialise the database cluster with initdb"
```
Won't dying here leave the password file behind?

sandeep-edb replied:
Die function will remove the password file if exists

dpage commented on server/scripts/windows/initcluster.ps1:
```diff
@@ -0,0 +1,321 @@
+    # on the entire path leading to the data directory
+    $arrDirs = $ParentOfDataDir.Split('\')
+    $nDirs = $arrDirs.Length - 1
+    Write-Host "`nNumber of directories: $nDirs"
```
Looks like debug code, probably not needed.

sandeep-edb replied:
yes, line 282 is a debug code. it's never used before so I agree it can probably be removed.

Manika-EDB replied:
removed that line.

dpage commented on server/scripts/windows/initcluster.ps1:
```diff
@@ -0,0 +1,321 @@
+    # Split the parent directory path into an array
+    $arrDirs = $ParentOfDataDir.Split('\')
+    $nDirs = $arrDirs.Length - 1
+    Write-Host "`nNumber of directories: $nDirs"
```
Looks like debug code, probably not needed.

sandeep-edb replied:
yes, line 164 is a debug code. it's never used before so I agree it can probably be removed.

Manika-EDB replied:
removed that line.

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
replacing getlocales.exe with installbuilder action

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB committed:
fix

Manika-EDB commented:
Hi @dpage, @sandeep-edb and @AKJUS 

@AKJUS Thank you for your approval!
@dpage I appreciate your review and have implemented the changes you suggested.

Additionally, I have made some modifications to improve support for locales containing non-ASCII characters:

1. Replaced `getlocales.exe` with a powershell command in `pgserver.xml.in` (https://github.com/EnterpriseDB/edb-installers/blob/dbp-719/initcluster/server/pgserver.xml.in#L2031-L2054)
2. Removed the code in `initcluster.ps1` that converted the locale to BCP-47 format (https://github.com/EnterpriseDB/edb-installers/blob/dbp-719/initcluster/server/scripts/windows/initcluster.ps1#L232-L238)
3. Since the locale is now passed to `initcluster.ps1` in BCP-47 formar (per point 1), the locale issue has been resolved.

Could you please review these changes and let me know if any further modifications are needed?

Thanks!

sandeep-edb commented:
LGTM

dpage reviewed (commented):
It looks fine to me in a quick eyeball review, but keep in mind it's been a long time since I wrote any InstallBuilder code, or did anything in PowerShell.

Manika-EDB committed:
fix

sandeep-edb committed:
Merge branch 'REL-17' into dbp-719/initcluster
